engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/stores/ExecutionStore.scala

```diff
@@ -0,0 +1,253 @@
+package cromwell.engine.workflow.lifecycle.execution.stores
+
+import common.collections.Table
+import cromwell.backend.BackendJobDescriptorKey
+import cromwell.core.ExecutionIndex.ExecutionIndex
+import cromwell.core.ExecutionStatus._
+import cromwell.core.{ExecutionIndex, JobKey}
+import cromwell.engine.workflow.lifecycle.execution.WorkflowExecutionActor.{apply => _}
+import cromwell.engine.workflow.lifecycle.execution.keys._
+import cromwell.engine.workflow.lifecycle.execution.stores.ExecutionStore._
+import wom.callable.ExecutableCallable
+import wom.graph.GraphNodePort.{ConditionalOutputPort, OutputPort, ScatterGathererPort}
+import wom.graph._
+import wom.graph.expression.ExpressionNode
+
+object ExecutionStore {
+  type StatusTable = Table[GraphNode, ExecutionIndex.ExecutionIndex, JobKey]
+
+  val MaxJobsToStartPerTick = 1000
+
+  implicit class EnhancedJobKey(val key: JobKey) extends AnyVal {
+    /**
+      * Given a StatusStable, return true if all dependencies of this key are in the table (and therefore are in this status),
+      * false otherwise.
+      */
+    def allDependenciesAreIn(statusTable: Table[GraphNode, ExecutionIndex.ExecutionIndex, JobKey]) = {
+      def portIsInStatus(outputPort: OutputPort) = {
+        // If the port is a scatter gather port, then it depends on the collector node which is stored at index None
+        val index = outputPort match {
+          case _: ScatterGathererPort => None
+          case _ => key.index
+        }
+
+        outputPort.executionNode.isInStatus(index, statusTable)
+      }
+
+      key match {
+        // For a scatter collector key, the dependencies are all the shards
+        case scatterCollector: ScatterCollectorKey => statusTable.row(scatterCollector.node).size == scatterCollector.scatterWidth
+        case conditionalCollector: ConditionalCollectorKey => statusTable.contains(conditionalCollector.node.executionNode, key.index)
+        // In the general case, the dependencies are held by the upstreamPorts
+        case _ => key.node.upstreamPorts forall portIsInStatus
+      }
+    }
+  }
+  
+  implicit class EnhancedOutputPort(val outputPort: OutputPort) extends AnyVal {
```

Comment:
could this be tacked on to `OutputPort`?

Reply:
Definitely could, I would say it depends on if we consider it being an artifact of how Cromwell execute WOM workflows in which case I'd argue it should be in Cromwell, or a WOM concept in which case putting it in the OutputPort class makes more sense.